### PR TITLE
Map pattern, creation, and update (partially)

### LINKED ISSUE #102: Support map pattern

### LINKED ISSUE #122: Support map-related expression
`#{a => b}` and `Map#{a => b}`

### PR DESCRIPTION
**Pull request by amutake**
issue: #102
To full support for map pattern, it is required to resolve #122.

### CODE REVIEW

File: lib/ast.ml [hunk 1]
```diff
@@ -19,6 +19,7 @@ type t =
     | MFA of {module_name: t; function_name: t; arity: t}
     | ListCons of t * t
     | ListNil
+    | Map of (t * t) list
```
[the remark this answers is not in the document]
Comment (yoshihiro503) [reply lib/ast.ml]: I understand that the difference is handled #122.

[the remark this answers is not in the document]
Comment (amutake) [reply lib/ast.ml]: In construction of map, only `#{A => B}` style is allowed. 

```
1> #{a := b}.
* 1: only association operators '=>' are allowed in map construction
```

In update of map, `M#{A := B}` and `M#{A => B}` are allowed, and I will handle this in #122.

**Comment by amutake** [comment]
rebased.

**Comment by amutake** [comment]
changed: I temporary considered all map as `map()`.

File: lib/ast.ml [hunk 2]
```diff
@@ -23,6 +23,8 @@ type t =
     | MFA of {module_name: t; function_name: t; arity: t}
     | ListCons of t * t
     | ListNil
+    | MapCreation of (t * t) list                  (* #{k1 => v1, ...} *)
+    | MapUpdate of t * (t * t) list * (t * t) list (* M#{k1 => v1, ..., ke1 := ve1, ...} *)
```
Comment (yoshihiro503) [inline lib/ast.ml]: How about using the following record?

```suggestion
    | MapUpdate of {map: t; assoc_fields: (t * t) list; exact_fields: (t * t) list} (* M#{k1 => v1, ..., ke1 := ve1, ...} *)
```

Comment (amutake) [reply lib/ast.ml]: Thanks! fixed: https://github.com/dwango/fialyzer/pull/141/commits/8061409a76017f97a424af7e5438b6aa637cc0c1

**Comment by amutake** [comment]
rebased and updated `derivation-rules.saty`

**Comment by amutake** [comment]
Thank you for your review!
I'll rebase and merge.

**Pull request merged**